### PR TITLE
Pre-build Python bytecode

### PR DESCRIPTION
**Pull request by legoktm**
## Status

Ready for review

## Description of Changes

* Generate Python bytecode at install time, ensures they're all owned by root:root and that it doesn't need to be generated on the fly.

Refs <https://github.com/freedomofpress/securedrop-security/issues/84>.

## Testing

* [ ] Build new packages, try installing them, make sure nothing breaks. Then try a submission and log into the JI (should exercise enough code paths that the pyc isn't obviously broken)
  * One harmless warning: `/opt/venvs/.../sqlalchemy/orm/query.py:196 SyntaxWarning: "is not" with a literal. Did you mean "!="?`
## Deployment

Any special considerations for deployment? Not particularly.

## Checklist

- [x] Linting (`make lint`) and tests (`make test`) pass in the development container
- [ ] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass
- [ ] I have written a test plan and validated it for this PR
- [x] These changes do not require documentation

### CODE REVIEW

**Comment by legoktm** [comment]
I tested everything except for logging into the JI, because the onion service keeps timing out for me :/

**Comment by zenmonkeykstop** [comment]
@legoktm with this change, can we update the apparmor profile to make the __pycache__ dirs readonly for apache?

**Comment by legoktm** [comment]
> @legoktm with this change, can we update the apparmor profile to make the **pycache** dirs readonly for apache?

Done but untested.

File: install_files/ansible-base/roles/build-securedrop-app-code-deb-pkg/files/usr.sbin.apache2 [hunk 1]
```diff
@@ -109,8 +109,8 @@
   /usr/share/distro-info/debian.csv r,
   /usr/share/file/magic r,
   /usr/share/file/magic.mgc r,
-  /opt/venvs/securedrop-app-code/**/__pycache__/ rw,
-  /opt/venvs/securedrop-app-code/**/__pycache__/* rw,
+  /opt/venvs/securedrop-app-code/**/__pycache__/ r,
```
Comment (zenmonkeykstop) [inline install_files/ansible-base/roles/build-securedrop-app-code-deb-pkg/files/usr.sbin.apache2:112]: Basic application functionality unaffected by this change, :+1:

File: install_files/securedrop-app-code/debian/postinst [hunk 2]
```diff
@@ -38,27 +38,27 @@ database_migration() {
...
         if ! echo "$sqlite_tables" | grep -q journalists; then
             # This is a fresh install, run the full migrations.
-            ${SDBIN}/alembic upgrade head
+            sudo -u www-data ${SDBIN}/alembic upgrade head
         else
             # This is an old install, so mark that the schema is up to date
             # as of the version of migration 15ac9509fc68_init.py when we first
             # started version controlling the database.
-            ${SDBIN}/alembic stamp 15ac9509fc68
+            sudo -u www-data ${SDBIN}/alembic stamp 15ac9509fc68
```
Comment (zenmonkeykstop) [inline install_files/securedrop-app-code/debian/postinst]: running `dpkg-reconfigure` to hit this line in the postinst regenerates `/var/www/securedrop/__pycache__/config.cpython-38.pyc` with `www-data` the owner.

If I run a downgrade and then upgrade again, both as `www-data`, it also recompiles config.py.

File: install_files/securedrop-app-code/debian/postinst [hunk 3]
```diff
@@ -97,11 +97,16 @@ adjust_wsgi_configuration() {
...
-# that obsolete dependencies can't linger after they've been removed.
 #
-remove_bytecode() {
+refresh_bytecode() {
+    # Remove any existing byte code files, to ensure that obsolete
+    # dependencies can't linger after they've been removed.
     find "${SDVE}" -name '*.py[co]' -delete
+    find /var/www/securedrop -name '*.py[co]' -delete
+    # Now generate the new byte-code
+    py3compile "${SDVE}"
```
Comment (zenmonkeykstop) [inline install_files/securedrop-app-code/debian/postinst:107]: bytecode is regenerated with root ownership in default locations :+1:

File: install_files/securedrop-app-code/debian/postinst [hunk 4]
```diff
@@ -205,7 +213,8 @@ case "$1" in
     service apache2 stop
 
     # Point Apache to our bundled mod_wsgi
-    ${SDBIN}/mod_wsgi-express module-config > /etc/apache2/mods-available/wsgi.load
+    # shellcheck disable=SC2024
+    sudo -u www-data ${SDBIN}/mod_wsgi-express module-config > /etc/apache2/mods-available/wsgi.load
```
Comment (zenmonkeykstop) [inline install_files/securedrop-app-code/debian/postinst]: is this necessary?

**Review by zenmonkeykstop — DISMISSED** [review]
Running the alembic tasks as www-data results in at least config.py being recompiled with `www-data` ownership. It might be worth dropping the "sudo -u www-data" for these and relying on existing logic to fix up ownership on the db after root does its thing.

**Comment by legoktm** [comment]
I dropped the `sudo -u www-data` part per review, I'll open a separate PR with that after some more testing.

**Review by zenmonkeykstop — APPROVED** [review]
LGTM, no impact on functionality and .pyc files are root-owned and not overwritten by Apache during a test source and journalist session.

**Pull request merged**